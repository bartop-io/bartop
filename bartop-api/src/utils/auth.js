const jwt = require('express-jwt');
const jwksRsa = require('jwks-rsa');
const config = require('../../config');

const checkJwt = jwt({
  // Dynamically provide a signing key based on the kid in the header and the
  // signing keys provided by the JWKS endpoint.
  secret: jwksRsa.expressJwtSecret({
    cache: true,
    rateLimit: true,
    jwksRequestsPerMinute: 5,
    jwksUri: `https://bartop.auth0.com/.well-known/jwks.json`
  }),

  // Don't require authorization in local development
  credentialsRequired: config.env !== 'local',

  // Validate the audience and the issuer.
  audience: config.auth.audience,
  issuer: `https://bartop.auth0.com/`,
  algorithms: ['RS256']
});

module.exports = checkJwt;
