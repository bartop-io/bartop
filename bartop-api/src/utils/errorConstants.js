module.exports = {
  UNAUTHORIZED: {
    message: 'Access token is missing or invalid.',
    code: 401
  },
  NONEXISTENT: {
    message: 'This route does not exist.',
    code: 404
  },
  CONTENT_TYPE: {
    message: 'Invalid content type.',
    code: 406
  },
  BODY_MODEL: {
    message: 'Invalid model found in body.',
    code: 422
  }
};
