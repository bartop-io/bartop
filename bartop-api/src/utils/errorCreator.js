module.exports.notFound = id => {
  const newError = new Error();
  newError.name = 'ResourceNotFoundError';
  newError.message = `Resource with ID:${id} does not exist.`;
  throw newError;
};
