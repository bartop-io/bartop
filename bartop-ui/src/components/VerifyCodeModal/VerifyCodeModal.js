import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Formik } from 'formik';
import { connect } from 'react-redux';
import { complement } from 'polished';

import Modal from '../Modal/Modal';
import CodeInput from '../CodeInput/CodeInput';
import {
  ModalContainer,
  ImageContainer,
  Image,
  Form
} from '../Modal/ModalComponents';

import strings from '../../strings';
import { actions as authActions } from '../../ducks/authentication/authentication';
import {
  actions as modalActions,
  MODAL_TYPES
} from '../../ducks/modals/modals';
import history from '../../singletons/history';
import VerifyCodeIcon from '../../images/verify-code-icon.svg';
import { colors } from '../styleUtils';

const VerifyImageContainer = ImageContainer.extend`
  background-color: ${complement(colors.lightBlueAccent)};
`;

const VerifyCodePrompt = styled.p`
  font-size: 20px;
  text-align: center;
`;

const MissingCodePrompt = styled.p`
  font-size: 14px;

  &:hover {
    cursor: pointer;
    text-decoration: underline;
  }
`;

export const VerifyCodeModal = ({ email, verifyCode, showModal, ...rest }) => (
  <Modal {...rest}>
    <Formik
      initialValues={{
        verificationCode: ''
      }}
      onSubmit={async (
        { verificationCode },
        { setSubmitting, setFieldError, setFieldValue }
      ) => {
        console.log('gonna subnmit', verificationCode);
        try {
          setFieldError('verificationCode', null);
          await verifyCode(email, verificationCode);
          setSubmitting(false);
          history.push({
            pathname: '/callback'
          });
        } catch (err) {
          console.error(err);
          const message =
            err.code === 'access_denied'
              ? strings.auth.verifyCodeNoMatch
              : strings.auth.verifyCodeFailureFallback;
          setSubmitting(false);
          setFieldError('verificationCode', message);
          // clear value on error, the previously entered code will be shown as the placeholder
          setFieldValue('verificationCode', '', false);
        }
      }}
      render={({
        values,
        errors,
        touched,
        handleChange,
        handleBlur,
        isSubmitting,
        setFieldValue,
        submitForm
      }) => (
        <ModalContainer>
          <VerifyImageContainer>
            <Image src={VerifyCodeIcon} />
          </VerifyImageContainer>
          <Form>
            <VerifyCodePrompt>{strings.auth.verifyPrompt}</VerifyCodePrompt>
            <CodeInput
              id="verificationCode"
              onChange={handleChange}
              onBlur={handleBlur}
              value={`${values.verificationCode}`}
              error={touched.verificationCode && errors.verificationCode}
              disabled={isSubmitting}
              submit={submitForm}
            />
            {/* if they didn't receive the code, allow them to go back a step while maintaining the email we sent the code to */}
            <MissingCodePrompt
              onClick={() =>
                showModal(MODAL_TYPES.LOGIN_MODAL, { prefillEmail: email })
              }
            >
              {strings.auth.missingCodePrompt}
            </MissingCodePrompt>
          </Form>
        </ModalContainer>
      )}
    />
  </Modal>
);

VerifyCodeModal.propTypes = {
  email: PropTypes.string.isRequired,
  verifyCode: PropTypes.func.isRequired,
  showModal: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  showModal: modalActions.showModal,
  verifyCode: authActions.verifyCode
};

export default connect(
  undefined,
  mapDispatchToProps
)(VerifyCodeModal);
