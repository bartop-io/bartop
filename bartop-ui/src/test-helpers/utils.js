export const noop = s => s;
