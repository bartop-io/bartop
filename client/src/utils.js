export const isExpired = expiresAt => {
  // check whether the current time is past the access token's expiry time
  return expiresAt <= new Date().getTime();
};
