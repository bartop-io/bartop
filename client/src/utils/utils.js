// determine when the token will expire
export const willExpireAt = expiresIn =>
  expiresIn * 1000 + new Date().getTime();

// check whether the current time is past the access token's expiry time
export const isExpired = expiresAt => expiresAt <= new Date().getTime();
