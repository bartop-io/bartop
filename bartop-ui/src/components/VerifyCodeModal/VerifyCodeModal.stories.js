import React from 'react';

import { storiesOf } from '@storybook/react';

import { VerifyCodeModal } from './VerifyCodeModal';
import { noop } from '../../test-helpers/utils';

const props = {
  verifyCode: noop,
  showModal: noop,
  email: 'billy@bartop.io',
  hideModal: noop,
  ariaHideApp: false
};

storiesOf('VerifyCodeModal', module).add('default', () => (
  <VerifyCodeModal {...props} />
));
