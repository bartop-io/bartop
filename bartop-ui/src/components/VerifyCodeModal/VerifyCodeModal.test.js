import React from 'react';
import { shallow } from 'enzyme';

import { VerifyCodeModal } from './VerifyCodeModal';
import { noop } from '../../test-helpers/utils';

const requiredProps = {
  email: 'bilbo@bartop.io',
  verifyCode: noop
};

it('matches the snapshot', () => {
  const wrapper = shallow(<VerifyCodeModal {...requiredProps} />);
  expect(wrapper).toMatchSnapshot();
});
