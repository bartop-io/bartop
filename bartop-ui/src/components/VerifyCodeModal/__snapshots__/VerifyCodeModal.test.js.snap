// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<Styled(Modal)>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "verificationCode": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  />
</Styled(Modal)>
`;
